## Importing necessary libraries

In [1]:
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 
%matplotlib inline
plt.style.use('ggplot')

## Getting Datsets from the drive

In [2]:
data=pd.read_csv("concrete_data.csv",sep=',',header=0)
data

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0      540.0                 0.0      0.0  162.0               2.5   
1      540.0                 0.0      0.0  162.0               2.5   
2      332.5               142.5      0.0  228.0               0.0   
3      332.5               142.5      0.0  228.0               0.0   
4      198.6               132.4      0.0  192.0               0.0   
...      ...                 ...      ...    ...               ...   
1025   276.4               116.0     90.3  179.6               8.9   
1026   322.2                 0.0    115.6  196.0              10.4   
1027   148.5               139.4    108.6  192.7               6.1   
1028   159.1               186.7      0.0  175.6              11.3   
1029   260.9               100.5     78.3  200.6               8.6   

      Coarse Aggregate  Fine Aggregate  Age  Strength  
0               1040.0           676.0   28     79.99  
1               1055.0           676.0   28     61.89  
2                932.0           594.0  270     40.27  
3                932.0           594.0  365     41.05  
4                978.4           825.5  360     44.30  
...                ...             ...  ...       ...  
1025             870.1           768.3   28     44.28  
1026             817.9           813.4   28     31.18  
1027             892.4           780.0   28     23.70  
1028             989.6           788.9   28     32.77  
1029             864.5           761.5   28     32.40  

[1030 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.

In [3]:
data.shape

(1030, 9)

In [4]:
data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
data.isnull().sum()  # no missing value in the dataset

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

In [6]:
# split the data set into predictors (inputs) and target (output)
predictors = data.drop(columns=['Strength'])
target = data['Strength']

n_cols = predictors.shape[1]

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
predictors.shape

(1030, 8)

In [9]:
n_cols=predictors.shape[1]
n_cols

8

In [10]:
target.shape

(1030,)

# Part B: Normalize the data

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [11]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [12]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

In [13]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [14]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [15]:
# Data Preparation -PartA
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3,random_state=42)

In [16]:
# build the model
model = regression_model()
model.fit(X_train,y_train,epochs=50,verbose=1)

Epoch 1/50
23/23 [==============================] - 1s 1ms/step - loss: 3090.5645
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 2001.0070
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1528.3304
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1143.7057
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 870.9944
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 716.5300
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 605.7163
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 533.7343
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 457.3549
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 415.7606
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 365.9110
Epoch 12/50
23/23 [==============================] - 0s 1ms/step - loss: 328.2994
Epoch 13/50
23/23 [==

In [17]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 2ms/step


107.26438903808594

In [18]:
from sklearn.metrics import mean_squared_error

In [19]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

107.26439195755303 0.0


### Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [20]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 93.40202331542969
10/10 [==============================] - 0s 2ms/step
MSE 2: 120.01332092285156
10/10 [==============================] - 0s 1ms/step
MSE 3: 111.53840637207031
10/10 [==============================] - 0s 1ms/step
MSE 4: 126.55146026611328
10/10 [==============================] - 0s 1ms/step
MSE 5: 116.00273132324219
10/10 [==============================] - 0s 1ms/step
MSE 6: 108.13201904296875
10/10 [==============================] - 0s 1ms/step
MSE 7: 116.37480926513672
10/10 [==============================] - 0s 1ms/step
MSE 8: 85.6135482788086
10/10 [==============================] - 0s 1ms/step
MSE 9: 98.19926452636719
10/10 [==============================] - 0s 2ms/step
MSE 10: 95.8898696899414
10/10 [==============================] - 0s 2ms/step
MSE 11: 76.58940887451172
10/10 [==============================] - 0s 1ms/step
MSE 12: 77.48164367675781
10/10 [==============================] - 0s 1ms/step
MSE 13: 87.79159545898438
10/10 [========================

    Peer-graded Assignment 	Mean MSE 	Standard Deviation
    A 	Build a baseline model 	57.64673 	 26.9559	
    B 	Normalize the data      65.75528 	 24.42921	
    C 	Increase the number of epochs(100) ? 	? 	
    D 	Increase the number of hidden layers (3) 	? 	?